eos-tech-support/eos-usb-internet-writer

```diff
@@ -0,0 +1,229 @@
+for app in "${apps_to_install[@]}"; do
+    flatpak install -y --or-update "$app"
+done
+
+# Prepare tmp directory
+mkdir -p "$tmp_dir/workspace"
+mkdir -p "$tmp_dir/downloads"
+
+# Download Software packages
+wget -c -O "$tmp_dir/downloads/kiwix-win.zip" http://download.kiwix.org/release/kiwix-desktop/kiwix-desktop_windows_x64.zip
+wget -c -O "$tmp_dir/downloads/kiwix-macos.dmg" https://download.kiwix.org/release/kiwix-desktop-macos/kiwix-desktop-macos.dmg
+
+# TODO: use real URLs when LE have the official builds
+wget -c -O "$tmp_dir/downloads/kolibri-win.zip" http://eor-desktop:8000/kolibri-win.zip
+wget -c -O "$tmp_dir/downloads/kolibri-macos.zip" http://eor-desktop:8000/kolibri-macos.zip
+
+# Copy Kiwix for Windows
+rm -rf "$tmp_dir/workspace/.kiwix-windows"
+unzip "$tmp_dir/downloads/kiwix-win.zip" -d "$tmp_dir/workspace"
+mv "$tmp_dir"/workspace/kiwix-desktop_windows* "$tmp_dir/workspace/.kiwix-windows"
+echo "$kiwix_library_xml" > "$tmp_dir/workspace/.kiwix-windows/library.xml"
+echo "$kiwix_windows_lnk_base64_data" | base64 -d > "$tmp_dir/workspace/Kiwix for Windows.lnk"
+
+# Copy Kiwix for macOS
+# TODO: Add support for auto loading ZIM files from USB key
+rm -rf "$tmp_dir/workspace/Kiwix for macOS.app"
+mkdir -p "$tmp_dir/downloads/kiwix-macos"
+umount "$tmp_dir/downloads/kiwix-macos" || true
+apfs-fuse "$tmp_dir/downloads/kiwix-macos.dmg" "$tmp_dir/downloads/kiwix-macos"
+cp -Lr "$tmp_dir/downloads/kiwix-macos/root/Kiwix.app" "$tmp_dir/workspace/Kiwix for macOS.app"
+umount "$tmp_dir/downloads/kiwix-macos"
+
+# Copy Kolibri for Windows
+rm -rf "$tmp_dir/workspace/.kolibri-windows"
+unzip "$tmp_dir/downloads/kolibri-win.zip" -d "$tmp_dir/workspace"
+mv "$tmp_dir/workspace/Kolibri" "$tmp_dir/workspace/.kolibri-windows"
+echo "$kolibri_windows_lnk_base64_data" | base64 -d > "$tmp_dir/workspace/Kolibri for Windows.lnk"
+
+# Copy Kolibri for macOS
+rm -rf "$tmp_dir/workspace/Kolibri for macOS.app"
+unzip "$tmp_dir/downloads/kolibri-macos.zip" -d "$tmp_dir/workspace"
+# Dereference symbolic links while renaming the package to make it look prettier
+cp -Lr "$tmp_dir/workspace/Kolibri.app" "$tmp_dir/workspace/Kolibri for macOS.app"
+rm -rf "$tmp_dir/workspace/Kolibri.app"
+
+# Copy ZIM files
+mkdir -p "$tmp_dir/workspace/.kiwix"
+cp "$(flatpak info -l com.endlessm.encyclopedia.en.Content)/files/com.endlessm.encyclopedia.en-openzim-subscription/1.zim" \
+   "$tmp_dir/workspace/.kiwix/wikipedia_en_endless_maxi.zim"
+
+# Setup Kolibri home
+# TODO: Install the actual content when ready
+run_kolibri manage importchannel network 9a46768f1ea24daa86ae693e4a04d04b
+run_kolibri manage importcontent network 9a46768f1ea24daa86ae693e4a04d04b
+
+# Write USB key
+extra_data=()
+while IFS= read -r -d '' path
+do
+    extra_data+=("--extra-data" "$path")
+done < <(find "$tmp_dir/workspace" -mindepth 1 -maxdepth 1 -print0)
+
+sudo $eos_write_live_image $debug "${extra_data[@]}" "$@"
```

Comment:
This script should probably be called with as root (or with `sudo`) instead of escalating privileges here, and we should check that similarly to how `eos-write-live-image` does it. Otherwise, if the calling user does not have admin access, we just wasted all the work up to this point.

Have you designed this so the `eos-write-live-image` arguments (like `--os-image`, `--persistent`, `--persistent-size` etc) are all passed after ` -- `? But it seems (unless I'm reading this wrong, which may be possible) you are also expecting a list of paths, which you then prepend each with `--extra-data`, is that it? If so, this seems a bit inconsistent and unexpected from the user's perspective. Also, whatever the approach is decided here, let's make it clear in the usage message.

(As a side note, I'm not a big fan of blindly forwarding command-line options like that in general, but I see the benefit of not having to parse and validate all of `eos-write-live-image` arguments, and keep them in sync between both scripts moving forward, so I think for the scope and intended usage of this tool this is the right approach.)

Reply:
I didn't want to run Kolibri and do the other processing stuff as root, I wasn't sure of the unseen implications of that so I played safe, maybe it's not that bad since I'm only sharing the Kolibri home folder to it.

That's exactly what it does, it prepends the `--extra-data` flags to copy the workspace filder and then it goes with everything after `--`, the flags after `--` could also include `--extra-data` if user wants, I could make it more clear in the usage message what's going on. Initially, my approach was to hardcode most of the flags passed to `eos-write-live-image` according to our USB specs and have a higher level flags for `eos-usb-internet-writer`. I decided later that this was maybe this was easier and more flexible

Reply:
Ok, makes sense not running the whole thing as root then. I wonder if there is a way we could check if the user has sudo access (I don't know if that is possible) in the beginning of the script. But I'm probably overthinking this, as the intended audience is Endless' devs. So yeah, let's keep this bit as is.

For the command line arguments, the direction you and @dylanmccall are discussing seems sensible to me.